Auto calibrate, pwmPassthrough, violent vibration

orbitinstasis opened the pull request:
Added auto calibration function.

This means the motor vibrates much more compared to before, note there have been a bunch of forum posts complaining about this drv2605 being 'too weak', this fixes that. 

Changed begin to take into account the new values found from the calibrator, and removed the no_overdrive.

Changed pwmPassthrough, fixing a bug and optimising it further.

orbitinstasis committed:
Example for using PWM Passthrough feature

demonstrates how pwm'ing a pin attached to IN/TRIG allows direct pwm control to your motor, via the DRV2605

orbitinstasis committed:
Deceleration for PWM Passthrough and PWMANALOG def

orbitinstasis committed:
PWMPassthrough function definition 

See the example on how to use this function, it allows any pwm pin connected to IN/TRIG to control the motor directly via the DRV2605

orbitinstasis committed:
Removed calibrate declaration

orbitinstasis committed:
Added note on changing the PWM Frequency

Have added change to pwm freq for teensy

orbitinstasis committed:
added declaration for auto calibration function

orbitinstasis committed:
Added auto-calibrator, and made output greater power

Added auto calibration function.

This means the motor vibrates much more compared to before, note there have been a bunch of forum posts complaining about this drv2605 being 'too weak', this fixes that. 

Changed begin to take into account the new values found from the calibrator, and removed the no_overdrive.

Changed pwmPassthrough, fixing a bug and optimising it further.

ryan-feeley commented:
Looks like a great commit! Can this be merged in?

ryan-feeley commented:
Follow-up on my comment above. The autoCalibration in this commit looks a little rough. I don't believe volatile is needed on goBit and I believe diag_results should be queried after polling the go bit for "0", not before.

orbitinstasis commented:
Thanks.

I'm a little new to the open source side of vcs, so I'm not entirely sure
what you mean by 'can it be merged', let me know if i should do something.

I will have a look at your follow up query later as I'm travelling today,
but iirc I followed the datasheet as strictly as I could. Have you tried
testing your changes? If it works differently / better then I can make the
changes and redo a pull request (can clean up autoCali routine).

Thanks



On 15 Nov 2017 07:42, "Ryan Feeley" <notifications@github.com> wrote:

Follow-up on my comment above. The autoCalibration in this commit looks a
little rough. I don't believe volatile is needed on goBit and I believe
diag_results should be queried after polling the go bit for "0", not before.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/adafruit/Adafruit_DRV2605_Library/pull/6#issuecomment-344510922>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AFg6c_86vMGE08aVs_bf5kO6nJ4gVtVgks5s2pXxgaJpZM4PX-NT>
.

ladyada commented:
@ryan-feeley wanna test/review this commit? that would help us merge it in faster :)

ryan-feeley commented:
@ladyada You've got it. I'll test and review tonight.

orbitinstasis commented:
Awesome. Let me know if there's anything you want me to change.

On 16 Nov 2017 00:51, "Ryan Feeley" <notifications@github.com> wrote:

> @ladyada <https://github.com/ladyada> You've got it. I'll test and review
> tonight.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adafruit/Adafruit_DRV2605_Library/pull/6#issuecomment-344779264>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFg6c1nNtFlLdbV0ncVohdUYNuH_nolwks5s24cogaJpZM4PX-NT>
> .
>

ryan-feeley commented:
I'm sad to report that I think we should leave this unmerged for the moment. I apologize that I didn't dig deeper before posting my initial comment that kicked this off. @orbitinstasis, if you'd like, I can issue a pull request on your fork to address my concerns, and once you've incorporated, you can re-issue your request here. I tried to figure out how to DM you, but I couldn't find any way to do that, so I'm just listing some highlights of the stuff I had trouble with below:

* The `.autoCalibrate()` method isn't implemented correctly. It uses `setMode(1)` as a proposed trigger of the auto-calibration. The value should be `setMode(0x07)`, or using the existing macro `setMode(DRV2605_MODE_AUTOCAL)`. It also sets the feedback register to 0xB6, which switches to a LFA motor.  Further, the `diag_results` variable needs to be set after polling the go bit for completion. Combined these issues cause the code to issue "calibration was successful" without that actually being the case.
* I don't think `.autoCalibrate()` should set the rated voltage levels (especially to non-reset values). These are application-specific and a client needs to have direct control over them. If we want to prevent incorrect use of auto-calibrate, perhaps we could put a flag in the private data that causes auto-calibrate to fail if the user hasn't explicitly configured the rated and overdrive voltages.
* I don't think we should call auto-calibrate directly in `.begin()`. When implemented correctly, the calibration routine will buzz the motor, and I think that is something a library client should control directly. Further, if the user has an LRA motor, they'll want to be able to do a sequence something like `.begin(), .useLRA(), .autoCalibrate()` which is not possible with the proposed implementation.
* I think we should eliminate the proposed `.pwmPassthrough()` method. This is effectively a convenience function that calls `Wire.begin(), self.autoCalibration(), self.setMode()`, and then zeros out the entirety control register3 (which tampers with several settings that apply more generally than for pwm activation). I think a better way to achieve this objective would be to define two new macros, DRV2605_MODE_PMW and DRV2605_MODE_ANALOG, and modify the existing `.setMode()` to handle these by setting the mode register to DRV2605_MODE_PWMANALOG (0x03) and bit1 of control register 3 to 0 for DRV2605_MODE_PMW and to 1 for DRV2605_MODE_ANALOG.

I apologize again for kicking this off and then backpedaling. @orbitinstasis is definitely taking this in the right direction by including auto-calibrate functionality. We just need to polish up the implementation. I expect the apparent benefit in output strength that this commit generates is realized by increasing the rated voltage level above the default, not from executing the calibration.

orbitinstasis commented:
Thanks for the feedback. The change in vibration is certainly through the
rated voltage as iirc I mentioned, it's currently set to output vcc to the
motor (but in reality has some drop likely through an internal diode).

I am on holiday at the moment and away from hardware, and it's been a while
since I added to this library so I'm going to take your word for what you
say,  and respond accordingly. I am glad you had a look at this, it's only
my second opensource project so I appreciate it very much.

RE 1: it seems changing the mode, feedback register, and order in which
diag result is set should fix the calibrate implementation. Do you have any
suggestions to what the feedback register should be set to?

RE 2: I originally intended to have the user set the rated voltage. It
could be implemented but I still don't understand what the overdrive
voltage does. Also, could you elaborate on 'especially not set to a non
reset value).

RE 3: again you make a good suggestion in respect to not calling AutoCal In
begin(), i had it the way you suggested but put the former in begin for
convenience.

RE 4: I will do a straight implementation of your suggestion here.

Thanks again for looking over this, and please don't worry about 'back
pedaling'!

I'll be back in the lab mid next week, at which I can begin to make the
changes.

Regarding vcs, please do what you think is best, and if you could then just
let me know what you're doing (and why, since I'm still a open source vcs
noob).

Thanks!

orbitinstasis commented:
Sorry for the massive radio silence. I've got the hardware and code back up and running and would like to implement @ryan-feeley 's suggestions so I can have my past attempts finally contributed, at least i can brag about helping adafruit in my CV. I saw your hardware all over UIST 2018!

Before i dig back into the code, can i ask if it'll still be ok to issue a pull request once i've worked this through, or is this now discontinued?

ladyada commented:
orbit, feel free to make a new PR!

orbitinstasis committed:
Making PR suggestions

correct use of setMode in auto cal.

Correct feedback register in Autocal.

polling diag_results after gobit completion. 

rated voltage and overdrive have default value from TI if not set, otherwise user can set them (0-255); using default: autoCalibration(), use client values: autoCalibration(ratedVoltage, overdrive);

moved auto-cal out of begin, and uncommented mode change at start of begin to get chip out of standby. Removed PWMPassthrough, not made suggestion yet for that.

orbitinstasis committed:
Update Adafruit_DRV2605.h

orbitinstasis commented:
Ok, i've made changes taking into account Ryans first three points. 

[I've updated my repo](https://github.com/orbitinstasis/Adafruit_DRV2605_Library) (not doing a PR yet) - i am getting a fail after attempting to auto calibrate when calling 'autoCalibration' after 'begin'; if i only call the autoCal function, it hangs. . It's worth noting TI have recently [updated the datasheet](http://www.ti.com/lit/ds/symlink/drv2605l.pdf), it seems more readable now.

I need to quit for the night, but i wouldn't mind someone having a look before i get deep into the datasheet again. On page 29, I am following the autoCal procedure and the only thing i think i'm not doing is populating all of the registers, which i think is fine since they'll be their default values.
